packages/neo-one-client-core/src/user/serializeUtils.ts

```diff
@@ -0,0 +1,263 @@
+export const serialize = (value: any): SerializedValueToken => {
+  if (value === null) {
+    return {
+      type: 'Null',
+    };
+  }
+  if (value === undefined) {
+    return {
+      type: 'Undefined',
+    };
+  }
+  if (typeof value === 'boolean') {
+    return {
+      type: 'Boolean',
+      value,
+    };
+  }
+  if (typeof value === 'number') {
+    return {
+      type: 'Number',
+      value,
+    };
+  }
+  if (value instanceof Buffer) {
+    return {
+      type: 'Buffer',
+      value: value.toString('hex'),
+    };
+  }
+  if (value instanceof Map) {
+    return serializeMap(value);
+  }
+  if (BN.isBN(value)) {
+    return {
+      type: 'BN',
+      value: value.toString(),
+    };
+  }
+  if (BigNumber.isBigNumber(value)) {
+    return {
+      type: 'BigNumber',
+      value: value.toString(),
+    };
+  }
+  if (typeof value === 'object') {
+    if (value instanceof Array) {
+      return serializeArray(value);
+    }
+    if (value instanceof Error) {
+      return {
+        type: 'Error',
+        value: value.message,
+      };
+    }
+    if (value.type === 'SourceMap') {
+      return {
+        type: 'SourceMap',
+        value: serializeObject(value.value),
+      };
+    }
+
+    return serializeObjectToken(value);
+  }
+  if (typeof value === 'string') {
+    return {
+      type: 'String',
+      value,
+    };
+  }
+
+  throw new Error(`Attempted to serialize invalid value: ${JSON.stringify(value)}`);
+};
+
+export const deserializeObject = (serializedObject: SerializedObject): DeserializedObject =>
+  Object.entries(serializedObject).reduce<DeserializedObject>(
+    (acc, [key, value]) => ({
+      ...acc,
+      [key]: deserialize(value),
+    }),
+    {},
+  );
+
+export const deserializeArray = (serializedArray: SerializedArray): DeserializedArray =>
+  serializedArray.map(deserialize);
+
+export const deserialize = (serializedValue: SerializedValueToken) => {
+  switch (serializedValue.type) {
+    case 'Undefined':
+      return undefined;
+    case 'Null':
+      // tslint:disable-next-line no-null-keyword
+      return null;
+    case 'String':
+      return serializedValue.value;
+    case 'Number':
+      return serializedValue.value;
+    case 'Boolean':
+      return serializedValue.value;
+    case 'BigNumber':
+      return new BigNumber(serializedValue.value);
+    case 'BN':
+      return new BN(serializedValue.value);
+    case 'Buffer':
+      return Buffer.from(serializedValue.value, 'hex');
+    case 'Object':
+      return deserializeObject(serializedValue.value);
+    case 'Array':
+      return deserializeArray(serializedValue.value);
+    case 'Map':
+      return deserializeArray(serializedValue.value);
+    case 'SourceMap':
+      return Promise.resolve(deserializeObject(serializedValue.value));
+    case 'Error':
+      return new Error(serializedValue.value);
+    default:
+      throw new Error(`Attempted to deserialize invalid value: ${JSON.stringify(serializedValue)}`);
```

Comment:
We typically do something like
```
      /* istanbul ignore next */
      utils.assertNever(type);
      /* istanbul ignore next */
      throw new Error('For TS');
```

The important bit being the `assertNever` part which ensures statically that you've hit every possible `type` value.